nixos/proxychains: init

sorki opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Modular configuration for `proxychains`. After enabling, it allows running `proxychains4` out of box proxying connections to Tor SOCKS when `services.tor.client` is enabled.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sorki committed:
nixos/proxychains: init

sorki commented:
Proxies default now reflects `services.tor.client.enable` correctly.

mmilata reviewed (approved):
Looks alright.

Lassulus commented:
hey, sorry for the delay.
new services are encouraged to have a test with them. Can you add one?

balsoft commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+
+    services.proxychains = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enable to install proxychains configuration..
+        '';
```
Why not use `mkEnableOption` ?

Infinisil commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+    ${cfg.tcpConnectTimeOut}
+    ${cfg.localnet}
+    ${cfg.proxies}
+    ${cfg.extraConfig}
```
Does it make sense to have an `extraConfig` at the end? From looking at the config file, it doesn't seem like anything other than proxies can go after the `[ProxyList]` section.

Infinisil commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+    ${cfg.extraConfig}
+  '';
+
+  proxyOptions = {
```
Looking at the [sample config file](https://github.com/haad/proxychains/blob/master/src/proxychains.conf) (which I think would be great to put in proxychains `$out`), you can also pass a password, which is missing here yet.

sorki replied:
Done via f9bdbf2a13d9db1f51f7b361cb3de7cb68bea9ab

sorki replied:
I've deliberately decided to not include `user` and `password` proxy options since it seems pretty obscure (the most common use-case is just running this with local Tor proxy) and poses a problem with passwords-in-store.

Infinisil commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+      port = mkOption {
+        type = types.port;
+        description = "Proxy port";
+        apply = builtins.toString;
```
Don't use `apply`, it prevents other modules from reusing module options values. Instead use `toString` where `port` is used. Same for all the other uses of `apply`.

Infinisil commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+      };
+
+      proxies = mkOption {
+        type = types.listOf (types.submodule proxyOptions);
```
Don't use `listOf`, because it doesn't allow changing definitions. Use `attrsOf` instead.

Infinisil commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+          '';
+      };
+
+      extraConfig = mkOption {
```
Ideally such an option wouldn't be added unless there's a good reason for it, see also NixOS/rfcs#42

Infinisil commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+        default = "";
+        description = ''
+          Extra configuration. Contents will be added verbatim to proxychains
+          configuration file.
```
And if it is added, it should point to where all the options are documented

Infinisil commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+      '';
+    };
+
+    services.proxychains.proxies = mkIf config.services.tor.client.enable
+      (mkDefault [{
+        type = "socks4";
```
I don't think this should be in this module, but rather in the tor one, and maybe behind an enable option

sorki replied:
Not sure how to call the option - `services.tor.client.proxychains.enable`?

Looks like additional pitfall when you want to use proxychains with Tor - the sample `proxychains.conf` has Tor proxy enabled as well and we only enable it when `tor.client` is enabled.

Infinisil commented on nixos/modules/services/security/proxychains.nix:
```diff
@@ -0,0 +1,184 @@
+        port = 9050;
+      }]);
+
+    environment.etc."proxychains.conf".text = configFile;
+    environment.systemPackages = [ pkgs.proxychains ];
```
`proxychain` is under `services.*`, but I don't see any service being defined. If it isn't a service it should be under `programs.*` instead.

sorki committed:
proxychains: copy src/proxychains.conf to /etc

As requested by Infinisil in #86225.

sorki commented:
- Fixed `mkEnableOption`
- Dropped `extraArgs`
- Switched to `programs.proxychains` instead of `services`
- `attrsOf` instead of `listOf` for `programs.proxychain.proxies`
- Removed option `apply`ies

Infinisil commented on nixos/modules/programs/proxychains.nix:
```diff
@@ -0,0 +1,161 @@
+  '';
+
+  proxyOptions = {
+    options = {
```
Another suggestion: Add an `enable` for each proxy such that they can easily be turned off without having to override the proxy attribute set.

Infinisil commented on nixos/modules/programs/proxychains.nix:
```diff
@@ -0,0 +1,161 @@
+        torproxy = {
+          type = "socks4";
+          host = "127.0.0.1";
+          port = 9050;
```
And this should probably be set like
```nix
mkIf config.services.tor.client.enable {
  torproxy = mkDefault {
    type = "socks4";
    host = "127.0.0.1";
    port = 9050;
  };
}
```

Otherwise specifying any other proxy would clear this one.

sorki commented:
Thanks for the suggestions Infinisil, also fixed the wording of the `programs.proxychains.enable` description to be rendered nicely in manual.

sorki commented:
I think this is good to go.

infinisil reviewed (approved):
Yup, looks good!
